enacts/calc.py

```diff
@@ -619,6 +620,103 @@ def daily_tobegroupedby_season(
     return daily_tobegroupedby_season
 
 
+def assign_season_coords(
+    time_series,
+    start_day,
+    start_month,
+    end_day,
+    end_month,
+    time_dim="T",
+):
+    """Assign season start and season end coordinates to this object,
+    mapping time dimension dates that belong to a season to their season start and end.
+    Data outside seasons are dropped.
+
+    Parameters
+    -----------
+    time_series : DataArray, Dataset
+        Time-dependent data to be grouped.
+    start_day : int
+        Day of the start date  of the season.
+    start_month : int
+        Month of the start date of the season.
+    end_day : int
+        Day of the end date of the season.
+    end_month : int
+        Day of the end date of the season.
+    time_dim : str, optional
+        Time coordinate in `time_series` (default `time_dim` ="T").
+
+    Returns
+    -------
+    time_series : DataArray, Dataset
+        `time_series` where days outside the seasons of interest are dropped
+        and with two additional coordinates `season_start` and `season_end`
+        that map the time dimension dates that belong to a season to
+        their season start and end respectively.
+
+    See Also
+    --------
+    xarray.DataArray.assign_coords, xarray.DataArray.groupby
+
+    Notes
+    -----
+    If the first (last) season is truncated, because `time_series` starts (ends) within it,
+    The first (last) element of `season_start` ( `season_end` ) differs from all others
+    and is the first (last) element of `time_dim` .
+    The additional coordinates are expected to be used by a groupby function.
+
+    Examples
+    --------
+    """
+    if (
+        (start_day == 29)
+        and (start_month == 2)
+        and ((~time_series[time_dim].dt.is_leap_year).sum() > 0)
+    ) :
+        raise Exception(
+            "if there is at least one non-leap year in time_coord, can not start on 29-Feb"
+        )
+    # ending season on 29-Feb means ending on 1-Mar with -1 day offset
+    if (end_day == 29 and end_month == 2) :
+        end_day = 1
+        end_month = 3
+        day_offset = -1
+    else:
+        day_offset = 0
```

Comment:
Maybe end_month/end_day should *always* represent the first day that's not included in the season? I.e. day_offset is always -1, then the DJF season would be represented with end_month = 3, end_month = 1, and JFM with end_month = 4, end_day = 1. Advantages: no special case --> less likelihood of a bug in the special case that we don't notice because we failed to test it; corresponds to e.g. how arrays are sliced in python: `'012345'[0:4]` == '0123'`, not `'01234'`. Disadvantage: for some reason xarray slicing, unlike python slicing, uses intervals that are closed on both ends, so that could be confusing.

Reply:
It's late in the review cycle for me to be introducing something like this. Feel free to ignore if you don't like the suggestion.

Reply:
I can see arguments for (or against) either way.

I am not sure there won't be any special case. If the user was a season ending in 28 Feb (included), would they put 29 Feb or 1 Mar? If they put 29 Feb, they would miss 3 28 Feb out of 4; if they put 1 Mar, they would include 29 Feb every 4 years. At least if using sel_day_and_month...

If we envision to function to work also for the monhtly case, then start/end_day should be an optional input and having its default value at 1 would be straightforward with open interval end. If we keep the intervals closed on both ends, we should probably set them to None, have start be 1, and have end be 1 too with an offset of -1 so that when users don't enter days, they get full months. So it's more gymnastics (but not crazy) in that case.

Then, as we try to build on top of xarray (well at least that is what I am trying to do), I feel we should try and opt xarray conventions, and thus keep closed intervals... That's not a very strong argument.

Reply:
> If they put 29 Feb, they would miss 3 28 Feb out of 4

I don't understand. Why would they miss Feb 28?

Xarray's choice to define slices as closed on both ends doesn't make sense to me. You can't partition a range of real values into bins using intervals that are closed on both ends. Indeed, [`groupby_bins`](https://docs.xarray.dev/en/stable/generated/xarray.DataArray.groupby_bins.html) defaults to intervals that are open on the left and closed on the right, which is doubly weird, because it's consistent with neither the behavior of `.sel` nor the more general convention, which (as I see it) is for intervals to be closed on the left and open on the right.

Reply:
They would miss it if we keep on using [sel_day_and_month](https://github.com/iridl/python-maprooms/blob/8e9d3021054eaf9a177e3dcd7e4e5e96a8d632e7/enacts/calc.py#L480)

Yes I noticed the seemingly strange behavior of groupby_bins with respect to intervals edges inclusivity. They just reproduced the [pandas.cut ](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.cut.html)behavior though. And maybe it is so because this is more about categorizing rather than slicing? Maybe we should not try, use and think through the groupby prism but through the [resample](https://docs.xarray.dev/en/stable/generated/xarray.DataArray.resample.html) one?